`Profile.fetch()` throwing on recent master

Specifically, this assertion is hit:

https://github.com/JuliaLang/julia/blob/c5f348726cebbe55e169d4d62225c2b1e587f497/stdlib/Profile/src/Profile.jl#L536

Reproducer is a little.. challenging:

 1) dev/clone master of https://github.com/Seelengrab/PropCheck.jl (I haven't checked whether other code also throws, sorry)
 2) run this:

```julia
using PropCheck
iExample = igen(UInt8(100))
l = PropCheck.vector(11, iExample)
PProf.@pprof check(issorted, l)
```

or just a regular `Profile` run around the last line instead of `@pprof`. This throws with something similar to

```julia
julia> PProf.@pprof check(issorted, l)
┌ Info: Found counterexample for 'issorted', beginning shrinking...
└   t = Tree(UInt8[0x3e, 0x0b, 0x4c, 0x4f, 0x28, 0x27, 0x3c, 0x5a, 0x1e, 0x3f, 0x24])
ERROR: AssertionError: metadata stripping failed i=35 j=0 data[1:i]=UInt64[0x00007f7271154ee8, 0x00007f727115addb, 0x00007f726c9865d2, 0x00007f726cb34d91, 0x00007f726ddd1f78, 0x00007f726cb34296, 0x00007f72707d575a, 0x00007f726e58b434, 0x00007f726e596560, 0x00007f726e56b664, 0x00007f726e56964d, 0x00007f726e569e18, 0x00007f726e572339, 0x00007f726e5731cc, 0x00007f726e5625ed, 0x00007f726e56a24e, 0x00007f726e56aa93, 0x00007f726e56ae75, 0x00007f726e56b26f, 0x00007f726e56b493, 0x00007f72707d85be, 0x00007f72707d983c, 0x00007f72707daef7, 0x00007f72707dbefc, 0x00007f7270748cfc, 0x00007f7270749c81, 0x00007f727076710b, 0x00007f7270766a01, 0x00007f7270767673, 0x00007f7270768666, 0x00007f7270786bf0, 0x00007f7270787819, 0x00007f7270788b41, 0x00007f725d5dbaf2, 0x00007f725d5dc134]
Stacktrace:
[1] fetch(; include_meta::Bool)
@ Profile ~/julia/usr/share/julia/stdlib/v1.8/Profile/src/Profile.jl:536                                                                                                                         
[2] fetch
@ ~/julia/usr/share/julia/stdlib/v1.8/Profile/src/Profile.jl:507 [inlined] 
[3] pprof(data::Nothing, lidict::Nothing; sampling_delay::Nothing, web::Bool, webhost::String, webport::Int64, out::String, from_c::Bool, full_signatures::Bool, drop_frames::Nothing, keep_frames::Nothing, ui_relative_percentages::Bool)
@ PProf ~/.julia/packages/PProf/QeQxG/src/PProf.jl:106                                                                                                                                           
[4] pprof (repeats 2 times)
 @ ~/.julia/packages/PProf/QeQxG/src/PProf.jl:105 [inlined]                                                                                                                                       
[5] top-level scope
@ ~/.julia/packages/PProf/QeQxG/src/PProf.jl:349
```

This should _not_ be dependent on `PProf`, so I think just

```julia
try                                                                                   
    var"#29#status" = Profile.start_timer()                                           
    if var"#29#status" < 0                                                            
        Profile.error(Profile.error_codes[var"#29#status"])                           
    end                                                                               
    check(issorted, l)                                                                
finally                                                                               
    Profile.stop_timer()                                                              
end                                                                                   
data = Profile.fetch()
```

should also throw.

1.7-rc1 works fine though, as well as using `Profile.fetch(;include_meta=true)`. CC @IanButterworth

ok, I can reproduce reliably on linux
```
julia> Profile.fetch();
ERROR: AssertionError: metadata stripping failed i=5 j=0 data[1:i]=UInt64[0x00007fa473c64374, 0x00007fa473405cc7, 0x00007fa47330faf3, 0x00007fa45ececf23, 0x00007fa45ed3cfae]
Stacktrace:
 [1] fetch(; include_meta::Bool)
   @ Profile ~/Documents/GitHub/julia/usr/share/julia/stdlib/v1.8/Profile/src/Profile.jl:537
 [2] fetch()
   @ Profile ~/Documents/GitHub/julia/usr/share/julia/stdlib/v1.8/Profile/src/Profile.jl:508
 [3] top-level scope
   @ REPL[7]:1
```

--- a/src/stackwalk.c
+++ b/src/stackwalk.c
@@ -102,6 +102,13 @@ static int jl_unw_stepn(bt_cursor_t *cursor, jl_bt_element_t *bt_data, size_t *b
                 // But sometimes the external unwinder doesn't check that.
                 have_more_frames = 0;
             }
+            if (return_ip == 0) {
+                // The return address is clearly wrong, and while the unwinder
+                // might try to continue (by popping another stack frame), that
+                // likely won't work well, and it'll confuse the stack frame
+                // separator detection logic (double-NULL).
+                have_more_frames = 0;
+            }
             if (skip > 0) {
                 skip--;
                 from_signal_handler = 0;
@@ -141,8 +148,9 @@ static int jl_unw_stepn(bt_cursor_t *cursor, jl_bt_element_t *bt_data, size_t *b
             if (!from_signal_handler)
                 call_ip -= 1; // normal frame
             from_signal_handler = 0;
-            if (call_ip == JL_BT_NON_PTR_ENTRY) {
+            if (call_ip == JL_BT_NON_PTR_ENTRY || call_ip == 0) {
                 // Never leave special marker in the bt data as it can corrupt the GC.
+                have_more_frames = 0;
                 call_ip = 0;
             }
             jl_bt_element_t *bt_entry = bt_data + n;

--- a/stdlib/Profile/src/Profile.jl
+++ b/stdlib/Profile/src/Profile.jl
@@ -291,9 +291,12 @@ end
 
 function is_block_end(data, i)
     i < nmeta + 1 && return false
-    # 32-bit linux has been seen to have rogue NULL ips, so we use two to indicate block end, where the 2nd is the
-    # actual end index
-    return data[i] == 0 && data[i - 1] == 0
+    # 32-bit linux has been seen to have rogue NULL ips, so we use two to
+    # indicate block end, where the 2nd is the actual end index.
+    # and we could have (though very unlikely):
+    # 1:<stack><metadata><null><null><NULL><metadata><null><null>:end
+    # and we want to ignore the triple NULL (which is an ip).
+    return data[i] == 0 && data[i - 1] == 0 && data[i - 2] != 0
 end
 
 """

